core/src/main/java/org/elasticsearch/gateway/TransportNodesListGatewayStartedShards.java

```diff
@@ -110,47 +120,58 @@ protected NodesGatewayStartedShards newResponse(Request request, AtomicReference
+            failures.toArray(new FailedNodeException[failures.size()]));
     }
 
     @Override
     protected NodeGatewayStartedShards nodeOperation(NodeRequest request) {
         try {
             final ShardId shardId = request.getShardId();
-            final String indexUUID = request.getIndexUUID();
             logger.trace("{} loading local shard state info", shardId);
-            ShardStateMetaData shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger, nodeEnv.availableShardPaths(request.shardId));
+            ShardStateMetaData shardStateMetaData = ShardStateMetaData.FORMAT.loadLatestState(logger,
+                nodeEnv.availableShardPaths(request.shardId));
             if (shardStateMetaData != null) {
-                final IndexMetaData metaData = clusterService.state().metaData().index(shardId.getIndex()); // it's a mystery why this is sometimes null
-                if (metaData != null) {
-                    ShardPath shardPath = null;
-                    try {
-                        IndexSettings indexSettings = new IndexSettings(metaData, settings);
-                        shardPath = ShardPath.loadShardPath(logger, nodeEnv, shardId, indexSettings);
-                        if (shardPath == null) {
-                            throw new IllegalStateException(shardId + " no shard path found");
-                        }
-                        Store.tryOpenIndex(shardPath.resolveIndex(), shardId);
-                    } catch (Exception exception) {
-                        logger.trace("{} can't open index for shard [{}] in path [{}]", exception, shardId, shardStateMetaData, (shardPath != null) ? shardPath.resolveIndex() : "");
-                        String allocationId = shardStateMetaData.allocationId != null ? shardStateMetaData.allocationId.getId() : null;
-                        return new NodeGatewayStartedShards(clusterService.localNode(), shardStateMetaData.legacyVersion, allocationId, shardStateMetaData.primary, exception);
-                    }
+                IndexMetaData metaData = clusterService.state().metaData().index(shardId.getIndex());
+                if (metaData == null) {
+                    // we may send this requests while processing the cluster state that recovered the index
+                    // sometimes the request comes in before the local node processed that cluster state
+                    // in such cases we can load it from disk
+                    metaData = IndexMetaData.FORMAT.loadLatestState(logger, nodeEnv.indexPaths(shardId.getIndex()));
                 }
-                // old shard metadata doesn't have the actual index UUID so we need to check if the actual uuid in the metadata
-                // is equal to IndexMetaData.INDEX_UUID_NA_VALUE otherwise this shard doesn't belong to the requested index.
-                if (indexUUID.equals(shardStateMetaData.indexUUID) == false
```

Comment:
yes, the folders are renamed before we open the transport gates. We now verify the uuid in the cluster state metadata and/or read the meta data file from uuid'ed folder which I think make this test redundant. On top of it, if we get it wrong, the allocation id will not match what the master is looking for.